pkg/operator/ceohelpers/bootstrap_test.go

```diff
@@ -0,0 +1,362 @@
+package ceohelpers
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+
+	"github.com/openshift/cluster-etcd-operator/pkg/operator/operatorclient"
+)
+
+func Test_GetBootstrapScalingStrategy(t *testing.T) {
+	tests := map[string]struct {
+		namespace      *corev1.Namespace
+		operatorState  operatorState
+		expectStrategy BootstrapScalingStrategy
+	}{
+		"default should be HA": {
+			namespace:      newNamespace(operatorclient.TargetNamespace).Build(),
+			operatorState:  newOperatorState().Build(),
+			expectStrategy: HAScalingStrategy,
+		},
+		"unsupported": {
+			namespace:      newNamespace(operatorclient.TargetNamespace).Build(),
+			operatorState:  newOperatorState().WithUnsupportedModeEnabled().Build(),
+			expectStrategy: UnsafeScalingStrategy,
+		},
+		"non-ha": {
+			namespace:      newNamespace(operatorclient.TargetNamespace).WithNonHAAnnotation().Build(),
+			operatorState:  newOperatorState().Build(),
+			expectStrategy: NonHAScalingStrategy,
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			indexer := cache.NewIndexer(cache.MetaNamespaceKeyFunc, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc})
+			if test.namespace != nil {
+				if err := indexer.Add(test.namespace); err != nil {
+					t.Fatal(err)
+				}
+			}
+			fakeNamespaceMapLister := corev1listers.NewNamespaceLister(indexer)
+			fakeStaticPodClient := v1helpers.NewFakeStaticPodOperatorClient(test.operatorState.Spec, test.operatorState.Status, nil, nil)
+			actualStrategy, err := GetBootstrapScalingStrategy(fakeStaticPodClient, fakeNamespaceMapLister)
+			if err != nil {
+				t.Errorf("unexpected error: %s", err)
+				return
+			}
+			if test.expectStrategy != actualStrategy {
+				t.Errorf("expected stategy=%v, got %v", test.expectStrategy, actualStrategy)
+			}
+		})
+	}
+}
+
+func Test_IsBootstrapComplete(t *testing.T) {
+	tests := map[string]struct {
+		bootstrapConfigMap *corev1.ConfigMap
+		operatorState      operatorState
+		expectComplete     bool
+		expectError        error
+	}{
+		"bootstrap complete, nodes up to date": {
+			bootstrapConfigMap: newBootstrapConfigMap().
+				WithCompleteStatus().
+				Build(),
+			operatorState: newOperatorState().
```

Comment:
(No action required) The struct-based fixture creation makes more sense here, but I don't see it as being easy to determine your intent. Maybe define the various operator states used for testing as vars whose name would make that intent clear?

```golang
var (
  twoNodesAtCurrentRevision operatorState = ...
  twoNodesNotAllAtCurrentRevision operatorState = ...
  ...
```

Reply:
This is a good insight, I want to do some refactoring here.

Reply:
Okay, I decided to run with this and ended up removing all the builders and found what I think is a clear way to use static fixtures declared up top. I think the new factoring is in line with your suggestions and is far more readable now overall, but let's see if you and others agree! 😄

Reply:
Nice work, your intent is now very much clear!